packages/graphql-language-server/src/GraphQlLanguageServer.ts

```diff
@@ -0,0 +1,243 @@
+import {
+  InitializeResult,
+  TextDocumentChangeEvent,
+  InitializeParams,
+  IConnection,
+  TextDocumentPositionParams,
+  CompletionItem,
+  Range,
+  TextEdit,
+  CompletionItemKind,
+  TextDocument,
+  Diagnostic,
+  DiagnosticSeverity,
+} from 'vscode-languageserver';
+import uniqBy from 'lodash.uniqby';
+import {
+  StandardGraphQlParser,
+  StardogGraphQlParser,
+} from 'millan/dist/standalone/millan.graphql';
+import { autoBindMethods } from 'class-autobind-decorator';
+import {
+  errorMessageProvider,
+  AbstractLanguageServer,
+  CompletionCandidate,
+} from 'stardog-language-utils';
+import { ISemanticError } from 'millan';
+
+const SPARQL_ERROR_PREFIX = 'SPARQL Error: ';
+
+@autoBindMethods
+export class GraphQlLanguageServer extends AbstractLanguageServer<
+  StardogGraphQlParser | StandardGraphQlParser
+> {
+  protected parser: StardogGraphQlParser | StandardGraphQlParser;
+
+  constructor(connection: IConnection) {
+    // Like the SPARQL server, the GraphQl server instantiates a different parser
+    // depending on initialization params
+    super(connection, null);
+  }
+
+  onInitialization(params: InitializeParams): InitializeResult {
+    this.connection.onCompletion(this.handleCompletion);
+
+    if (
+      params.initializationOptions &&
+      params.initializationOptions.grammar === 'standard'
+    ) {
+      this.parser = new StandardGraphQlParser({
+        config: { errorMessageProvider },
+      });
+    } else {
+      this.parser = new StardogGraphQlParser({
+        config: { errorMessageProvider },
+      });
+    }
+
+    return {
+      capabilities: {
+        // Tell the client that the server works in NONE text document sync mode
+        textDocumentSync: this.documents.syncKind[0],
+        completionProvider: {
+          triggerCharacters: ['$'],
```

Comment:
No, it just guarantees that autocompletion _is_ triggered in that particular case. The Language Server Protocol spec says that it is generally up to editors to trigger autocompletion whenever necessary (e.g., when a character that could be the start of an identifier is entered), but this can be used to "force" triggering of autocompletion for characters that the editor wouldn't otherwise trigger on (e.g., characters that are invalid in identifiers).

Reply:
I see, thanks.